### DISCUSSION TITLE
Can I contribute to the project?

### DISCUSSION THREAD

**Discussion by songpola** [buijs-dev/klutter#14] [Q&A]
I used to be an Android dev and I love Kotlin. Now I am a Flutter dev but I hate Dart which lacks a lot of features compared to Kotlin.

I came across this project and it's what I'm looking for! I see you're doing this alone so I would appreciate it to help you accomplish the project roadmap if you allowed me to.

**Comment by buijs-dev** [comment]
Thanks for the offer! Glad to hear your excitement.

I'm curious to your reasoning why you hate Dart? And also why are you a Flutter developer now despite said hate for Dart? This project does require a fair amount of Dart coding and possibly (probably) a lot more in the future (see for example the 3 dart packages that are used by klutter: https://pub.dev/publishers/buijs.dev/packages.

As for collaboration... If you are seriously interested then it would be best to contribute by working on a feature and opening a PR. For that to work you'd need a proper roadmap and feature/story descriptions. Due to this being a hobby project, those are not available (yet). 

I am currently wrapping up a major release (2023.1.1.beta) and if I find some time I can for sure set out a roadmap and describe some features.

**Reply by songpola** [reply]
Sorry for late reply. Here's my answer:

> Why I hate Dart?

Actually, it's not that I hate Dart, but rather I find it inconvenience for me, a developer that came from Kotlin, as Dart lacks lots of QoL features as the way language was designed to be [*familiar*](https://dart.dev/guides/language/effective-dart#:~:text=Be%20brief.,simply%20and%20easily.).

For example, I always faced many problem when it comes to dealing with data in the app. I just want a simple data classs to store some simple state. In Dart, I need a 3rd-party packages, like `freezed` (and `build_runner`) or `built_value` or `equatable`, and a lot of boilerplate and generated files, just for simple development-purposed feature. On the other hand. In Kotlin, all you need is just put a `data` keyword in front of the class declaration and... Vola! (We're not talking about state management. That's another one)

But Dart also has pros that Kotlin doesn't, like AOT-JIT compilation (Hot-reload) which is very very useful for frontend UI development. (I think that's why Flutter use Dart in the first place though)

So, Dart and Kotlin both have points they're good at. Still, if someone ask me whether I prefer Dart or Kotlin. I would instantly reply, "KOTLIN ❤️". Because ~80% of my time was spent dealing with business logic in the app, not the UI.

> Why I'm Flutter developer now?

Well, I switched to Flutter because of many reasons, like:
1. It is so easy to make a UI with declarative approach.
2. Write once, run anywhere! 😂
and many more.

But I always keep looking for a way to mix-and-match the technologies that could possibly works together.
One of my dream is to use Dart/Kotlin (Using Dart for UI and do all the business logic in Kotlin).
Which I think it's possible! Considering Dart have `ffi`, `jni` and JS-interop which is in active development. Kotlin also have their Multiplatform and WASM target coming this year (or very soon) and even K2.
For me, I foresee a bright future for both languages.
(I actually have a case-study for Flutter(+Tauri)/Rust: [AppFlowy](https://github.com/AppFlowy-IO/AppFlowy))

And when I came accross this project, it's like one step closer to my dream. That's why I'm so hyped!

Can't wait to see your plan on this project and I will try to fully understand this project for contributions!

**Reply by buijs-dev** [reply]
Well what you're describing is definitely possible with klutter already. Write all the logic in the platform module (Kotlin) and use dart only for the UI code. 2023.1.1.beta will introduce streaming widget which will make that even easier!

Next step I plan on is writing UI using Kotlin Script, which can be pushed from the platform side to dart. So basically something akin to https://pub.dev/packages/json_dynamic_widget, but using Kotlin instead of JSON. That way you could write the full app in Kotlin and still enjoy the dart benefits.
